### PR TITLE
feat: recheck backfill sync

### LINKED ISSUE #9689: Recheck pipeline state after we finished pipeline sync

### PR DESCRIPTION
**Pull request by mattsse**
closes #9689

checks if we need to trigger another backfill run after pipeline finished

### CODE REVIEW

File: crates/engine/tree/src/tree/mod.rs [hunk 1]
```diff
@@ -409,6 +409,63 @@ where
...
+        // The block number that the backfill finished at - if the progress or newest
+        // finalized is None then we can't check the distance anyways.
+        //
+        // If both are Some, we perform another distance check and return the desired
+        // backfill target
+        let Some(backfill_target) =
+            ctrl.block_number().zip(newest_finalized).and_then(|(progress, finalized_number)| {
+                // Determines whether or not we should run backfill again, in case
+                // the new gap is still large enough and requires running backfill again
+                self.backfill_sync_target(progress, finalized_number, None)
```
Comment (onbjerg) [inline crates/engine/tree/src/tree/mod.rs:457]: what happens if the finalized block hash is different but the number is the same?

Comment (mattsse) [reply crates/engine/tree/src/tree/mod.rs:457]: different compared to what?

File: crates/engine/tree/src/tree/mod.rs [hunk 2]
```diff
@@ -409,6 +409,63 @@ where
...
+            .state
+            .buffer
+            .block(&sync_target_state.finalized_block_hash)
+            .map(|block| block.number);
+
+        // TODO: state housekeeping
+
+        if sync_target_state.finalized_block_hash.is_zero() {
+            return
+        }
```
Comment (fgimenez) [inline crates/engine/tree/src/tree/mod.rs:446]: Could we move this check before declaring `newest_finalized`?

**Review by fgimenez — APPROVED** [review]
nice

File: crates/engine/tree/src/tree/mod.rs (hunk 1 above)
Comment (onbjerg) [reply crates/engine/tree/src/tree/mod.rs:457]: the block on disk.

Comment (mattsse) [reply crates/engine/tree/src/tree/mod.rs:457]: we don't trigger backfill, only if the distance from synced progress to (buffered) finalized block is large enough

**Pull request merged**